### PR TITLE
Allow running StandaloneApiServer in read-only mode.

### LINKED ISSUE #5635: Remove ReadService from StandaloneApiServer
Currently, `StandaloneApiServer` reads the initial ledger config and the ledger ID from the `ReadService`. Both settings should be defined in a ledger API server config, and the ledger API server should not depend on the ReadService at all.

At the same time, remove `ReadService.inititalConditions` and ask all users to set the ledger ID and initial configuration through the ledger API server config.

**It's important to note the following:**

This will remove the only connection of `ledgerId` between the participant and the ledger backend implementation.
As a consequence, the ledger ID must always be provided via command line parameter/config, and all participant components validate the expected ledgerId against the one already stored in a persistent index.

**Indexer**
* If no ledger ID is found in persistent storage, store the ledger ID provided via CLI/config. 
* Otherwise if the ledger ID provided via CLI/config does not match the ledger ID in persistent storage, abort initialization.

**Ledger API Server**
1. If no ledger ID is found in persistent storage, keep polling persistent storage with exponential back off until the ledger ID can be read.
2. If the ledger ID provided via CLI/config  does not match the ledger ID in persistent storage, abort initialization.

### PR DESCRIPTION
**Pull request by gerolf-da**
Also fixes #5635.

CHANGELOG_BEGIN
[DAML Integration Kit]: ``StandaloneApiServer`` can now be run in a read-only mode.
  - The type of the constructor parameter ``writeService`` of ``StandaloneApiServer`` changed to ``Option[WriteService]``. Passing ``None`` will not start any of the admin services, the command service, and the command submission service.
  - The constructor parameter ``readService`` of ``StandaloneApiServer`` has been removed.
  - A new constructor parameter ``ledgerId`` has been added to ``StandaloneApiServer``. It is used to verify that that ``StandaloneApiServer`` is run against an index storage for the same ledgerId. Initialization is aborted if this is not the case.
CHANGELOG_END

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by gerolf-da** [comment]
There is currently no test for this and we probably should also add a `ReadOnlyRunner` to `kvutils/app`. WDYT?

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/index/ReadOnlySqlLedger.scala [hunk 1]
```diff
@@ -67,7 +67,12 @@ object ReadOnlySqlLedger {
             Future.failed(
               new LedgerIdMismatchException(foundLedgerId, initialLedgerId) with StartupException)
           case None =>
-            Future.successful(initialLedgerId)
+            logger.info("Ledger ID not found in the index database. Retrying again in 5 seconds.")
+            val promise = Promise[LedgerId]()
+            mat.scheduleOnce(5.seconds, () => {
+              promise.completeWith(verifyLedgerId(ledgerDao, initialLedgerId))
+            })
+            promise.future
```
Comment (stefanobaghino-da) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/index/ReadOnlySqlLedger.scala:75]: Does it make sense to handle the retry at this level? Would it make more sense to throw an exception here and let the caller decide what to do with this?

File: ledger/participant-state/kvutils/app/src/main/scala/com/daml/ledger/participant/state/kvutils/app/Config.scala [hunk 2]
```diff
@@ -123,9 +124,9 @@ object Config {
           config.copy(participants = config.participants :+ partConfig)
         })
       opt[String]("ledger-id")
-        .text(
-          "The ID of the ledger. This must be the same each time the ledger is started. Defaults to a random UUID.")
-        .action((ledgerId, config) => config.copy(ledgerId = Some(ledgerId)))
+        .optional()
```
[the remark this answers is not in the document]
Comment (gerolf-da) [reply ledger/participant-state/kvutils/app/src/main/scala/com/daml/ledger/participant/state/kvutils/app/Config.scala:127]: No, options are optional by default, but since other parameters had `optional()` explicit, I added it here as well.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/index/ReadOnlySqlLedger.scala (hunk 1 above)
Comment (gerolf-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/index/ReadOnlySqlLedger.scala:75]: So the question is how we control the retry mechanism. At this point, we know fairly well what went "wrong".

Alternatively we could slap a retry around the entire index initialization process and not really care about specific errors.

What do you think?

Comment (stefanobaghino-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/index/ReadOnlySqlLedger.scala:75]: I'm only a bit concerned of ending up with retry logic scattered across many classes. We can still signal a specific failure using a specific exception and handle it appropriately. Feel free to leave this here if you think it's better to deal with failures as close as possible to the source.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/apiserver/StandaloneApiServer.scala [hunk 3]
```diff
@@ -69,33 +68,28 @@ final class StandaloneApiServer(
...
-      authorizer = new Authorizer(
-        () => java.time.Clock.systemUTC.instant(),
-        initialConditions.ledgerId,
-        participantId)
       indexService <- JdbcIndex
         .owner(
           ServerRole.ApiServer,
-          initialConditions.config,
-          domain.LedgerId(initialConditions.ledgerId),
+          ledgerConfig.initialConfiguration,
```
Comment (rautenrieth-da) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/apiserver/StandaloneApiServer.scala:74]: This is a breaking change:
- the ledger ID is not given by the ledger backend (ReadService) anymore
- the initial time model is not given by the ledger backend (ReadService) anymore

I think the part about the time model should be added to the changelog as well.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/apiserver/StandaloneApiServer.scala [hunk 4]
```diff
@@ -124,7 +118,7 @@ final class StandaloneApiServer(
     } yield {
       writePortFile(apiServer.port)
       logger.info(
-        s"Initialized API server version ${BuildInfo.Version} with ledger-id = ${initialConditions.ledgerId}, port = ${apiServer.port}, dar file = ${config.archiveFiles}")
+        s"Initialized API server version ${BuildInfo.Version} with ledger-id = $ledgerId, port = ${apiServer.port}, dar file = ${config.archiveFiles}")
```
Comment (rautenrieth-da) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/apiserver/StandaloneApiServer.scala:121]: Not part of this PR, but should we also print more details at a lower log level? Maybe the whole `(ledgerId, config, commandConfig, partyConfig, ledgerConfig)` at DEBUG?

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/index/ReadOnlySqlLedger.scala (hunk 1 above)
Comment (rautenrieth-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/index/ReadOnlySqlLedger.scala:75]: If we do decide to leave the retry here, should we stop retrying at some point?

Comment (gerolf-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/index/ReadOnlySqlLedger.scala:75]: @stefanobaghino-da: I'm open to suggestions on where initialization errors should be handled. I think this falls into a similar category as waiting for the DB to be available (see #6146). There we currently retry 600 times with a retry delay of 1 second. We can do something similar here.

Since this is only run on startup I think it makes sense to have some leniency during the startup sequence between a read-only api server and the indexer, so not failing the entire process.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/apiserver/StandaloneApiServer.scala (hunk 3 above)
Comment (gerolf-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/apiserver/StandaloneApiServer.scala:74]: I think we should actually rather try to fetch the configuration from the database. We have the infrastructure now in place. Both here and in `LedgerBackedIndexServer#getLedgerConfiguration`. What do you think?

Comment (rautenrieth-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/apiserver/StandaloneApiServer.scala]: Yes, `JdbcIndex` should not depend on the initial configuration.

The only component that should access the initial configuration is `LedgerConfigProvider`, which is responsible for writing the initial configuration to the ledger if there is none.
